WordPress-iOS-Shared/Core/OperationPools.swift

```diff
@@ -0,0 +1,286 @@
+        super.init { request, completion in
+            let task = session.downloadTaskWithRequest(request) {
+                location, response, error in
+
+                let imageLocation = Result(value: location, error: error, fallbackError: OperationPoolError.noDataAndNoError.nserror)
+                let image = imageLocation.flatMap(parser)
+                completion(image)
+            }
+            task.resume()
+        }
+    }
+}
+
+/// An OperationPool to download images.
+///
+/// Builds on NetworkDownloadPool and transforms a `NSURLRequest` into a
+/// `UIImage`.
+///
+/// If the downloaded data can't be turned into an image, it results in a
+/// `OperationPoolError.invalidImageData` error.
+///
+public class ImageDownloadPool: NetworkDownloadPool<UIImage> {
+
+    /// Constructs a new ImageDownloadPool.
+    ///
+    /// Optionally, you can specify the NSURLSession to use. By default, it uses
+    /// the shared session.
+    ///
+    /// - parameter session: The NSURLSession to use for the network request. By
+    /// default it uses `NSURLSession.sharedSession()`.
+    ///
+    public init(session: NSURLSession = NSURLSession.sharedSession()) {
+        let parser: Parser = { url in
+            do {
+                let data = try NSData(contentsOfURL: url, options: .DataReadingUncached)
+                if let image = UIImage(data: data) {
+                    return .success(image)
+                } else {
+                    return .failure(OperationPoolError.invalidImageData.nserror)
+                }
+            } catch {
+                return .failure(error as NSError)
+            }
+        }
+        super.init(parser: parser, session: session)
+    }
+}
+
+/// Errors related to OperationPool.
+/// 
+public enum OperationPoolError: Int, ErrorType {
+    /// The NSURLSession completion block was called and both `data` and `error`
+    /// are `nil`. It doesn't seem like something that _should_ not happen, but
+    /// it's not enforced by the compiler and the documentation doesn't confirm
+    /// that it won't happen.
+    ///
+    case noDataAndNoError
+
+    /// The data we received was not a valid image, or at least not something
+    /// that UIImage(data:) could process.
+    ///
+    case invalidImageData
+
+    /// A description of the error, in English.
+    ///
+    /// - important: This description is for debugging purposes only and should
+    /// not be shown to the user.
+    ///
+    public var description: String {
+        // Untranslated as these error shouldn't be shown to users
+        switch self {
+        case .noDataAndNoError:
+            return "Network response had no data and no error"
+        case .invalidImageData:
+            return "Couldn't generate a valid image with the returned data"
+        }
+    }
+
+    /// The error converted to a NSError.
+    ///
+    public var nserror: NSError {
```

Comment:
I think it's useful that this is `public`. All the existing OperationPool implementations use `NSError` as their error type. This means that any requester can get a Result with a NSError in it.

You want to check, for instance, if the error is `.invalidImageData`. You then have a `OperationPoolError` and a `NSError` with no way to compare them.